lgc/util/llpcGfxRegHelper.h

```diff
@@ -0,0 +1,145 @@
+/*
+ ***********************************************************************************************************************
+ *
+ *  Copyright (c) 2020 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+ /**
+  ***********************************************************************************************************************
+  * @file  llpcGfxRegHelper.h
+  * @brief LLPC header file: contains the definition of LLPC utility class GfxRegHelper.
+  ***********************************************************************************************************************
+  */
+
+#pragma once
+#include "llpcGfxRegHandler.h"
+#include "llpcTargetInfo.h"
+
+namespace lgc
+{
+
+using namespace llvm;
+
+// =====================================================================================================================
+// Struct maintains state for specific register
+// - pValue : Cache the value of the register which is being fetched already.
+// - isModified : Set to be true if the register is being settled with new value.
```

Comment:
It's unclear to me what 'state' means here. Is it a symbolic expression over some pre-defined graphics register, or register as in arbitrary SSA value?
What does it mean for a register to be settled with a value? Do you write to a hardware register, construct a new llvm value over a previously read register value?

Reply:
e.g. if we want to get A2, and may keep the A2 if we did do some modification on it, and may want to use it in someplace else. 
```
D0 = extractelement(0) // DWORD0
M0 = 0x1111                 // mask
A1 =  M0 & D0              // specific register
A2 = A1 + 1                   // do some modification
D1 =(D0 & ~M0) | (A2 & M0) // update DWORD0
insertelement(0)                     // insert back
```

Reply:
I'd say that these are just symbolic expressions over values read from hardware graphics registers, materialized as llvm instructions. The code maintains a map from from a register ID to an llvm value with the current symbolic expression describing the updated state of the register.

The goal of the helper structs is to maintain the newest symbolic state of a register and eventually produce optimized llvm instruction sequences for that states.

If that sounds reasonable, could you add a similar description in comments?

Reply:
Thanks, i'll update the comment as you suggested.